### PR TITLE
Jwe

### PR DESCRIPTION
**Pull request by dgutt17**
- [Authentication - Session/Token Management](https://app.asana.com/0/1123302713673648/1200166249164848)

### CODE REVIEW

File: lib/lp_token_auth/core.rb [hunk 1]
```diff
@@ -1,4 +1,5 @@
 require 'jwt'
+require 'jwe'
```
Comment (bhennes2) [inline lib/lp_token_auth/core.rb:2]: Looks promising! Is there an additional `gem` dependency we need to include?

File: lib/lp_token_auth/core.rb [hunk 2]
```diff
@@ -21,19 +22,22 @@ def issue_token(id, **payload)
...
       end
 
-      JWT.encode(
+      jwt = JWT.encode(
         payload,
         LpTokenAuth.config.get_option(:secret),
         LpTokenAuth.config.get_option(:algorithm)
       )
+
+      JWE.encrypt(jwt, private_key, enc: 'A256GCM')
```
Comment (bhennes2) [inline lib/lp_token_auth/core.rb]: Would be nice is this was configurable via ENV var but defaults to `A256GCM`.

File: lib/lp_token_auth/core.rb (hunk 1 above)
Comment (dgutt17) [reply lib/lp_token_auth/core.rb:2]: We will need the [jwe gem](https://github.com/jwt/ruby-jwe). Its made by the same people that made the jwt gem.

File: lib/lp_token_auth/core.rb (hunk 2 above)
Comment (dgutt17) [reply lib/lp_token_auth/core.rb]: We can add that

**Comment by dgutt17** [comment]
@bhennes2 Do you think we should auto-generate private keys to store in the application.yml file so that devs won't have to do that manually?

File: lib/lp_token_auth/core.rb (hunk 2 above)
Comment (dgutt17) [reply lib/lp_token_auth/core.rb]: Now we can configure the encryption type. The env variable is called `JWE_ENCRYPTION`.

Comment (dgutt17) [reply lib/lp_token_auth/core.rb]: Could we also do this in the config file that gets made?

**Comment by chawes13** [comment]
@dgutt17 Is this considered a breaking change? If so, it'd be nice to include a migration guide

**Comment by dgutt17** [comment]
@chawes13 This probably will be a breaking change. I can create a migration guide for it as well. Would that be best in the Readme? Or is there another place we should put it?

**Comment by chawes13** [comment]
Typically we create another folder and link out in the README. [Decanter](https://github.com/LaunchPadLab/decanter) has an example of this!

**Comment by dschadd** [comment]
> Typically we create another folder and link out in the README. [Decanter](https://github.com/LaunchPadLab/decanter) has an example of this!

Perfect! I was just looking into the best way to create this. Thanks for linking.

**Comment by dgutt17** [comment]
@bhennes2 @dschadd I tried running the rake task in the `booster_rocket` to generate the private key but it didn't work.

@dschadd `booster_rocket` is the name of Empower's Rails API.

**Comment by dschadd** [comment]
> @bhennes2 @dschadd I tried running the rake task in the `booster_rocket` to generate the private key but it didn't work.
> 
> @dschadd `booster_rocket` is the name of Empower's Rails API.

What was the error?

**Comment by dgutt17** [comment]
`Don't know how to build task 'generate_rsa' (See the list of available tasks with rake --tasks)
/Users/dangutt/.rvm/gems/ruby-2.5.7/gems/rake-13.0.3/exe/rake:27:in <top (required)>'
/Users/dangutt/.rvm/gems/ruby-2.5.7/bin/ruby_executable_hooks:24:in eval'
/Users/dangutt/.rvm/gems/ruby-2.5.7/bin/ruby_executable_hooks:24:in <main>'`

**Comment by dschadd** [comment]
> `Don't know how to build task 'generate_rsa' (See the list of available tasks with rake --tasks) /Users/dangutt/.rvm/gems/ruby-2.5.7/gems/rake-13.0.3/exe/rake:27:in <top (required)>' /Users/dangutt/.rvm/gems/ruby-2.5.7/bin/ruby_executable_hooks:24:in eval' /Users/dangutt/.rvm/gems/ruby-2.5.7/bin/ruby_executable_hooks:24:in <main>'`

What did you run to get this error?

**Comment by dgutt17** [comment]
@dschadd I ran `rake generate_rsa` in the `booster_rocket` repo

File: lib/lp_token_auth/core.rb [hunk 3]
```diff
@@ -56,5 +60,23 @@ def check_id!(id)
         raise LpTokenAuth::Error, "id must be a string or integer, you provided #{id}"
       end
     end
+
+    private
+
+    def private_key
+      # raise LpTokenAuth::NoKeyPresent if ENV['JWE_PRIVATE_KEY'].blank?
```
Comment (bhennes2) [inline lib/lp_token_auth/core.rb]: @dgutt17 Can we remove this commented out code if the `begin.....rescue` is elsewhere?

File: lp_token_auth.gemspec [hunk 4]
```diff
@@ -16,6 +16,7 @@ Gem::Specification.new do |s|
   s.files                 = `git ls-files -z`.split("\x0").reject { |f| f.match(%r{^(test|spec|features)/}) }
 
   s.add_dependency        'jwt', '>= 1.5.6'
-  s.add_development_dependency 'rake', '~> 10.4', '>= 10.4.2'
+  s.add_dependency        'jwe', '~> 0.4.0'
```
Comment (bhennes2) [inline lp_token_auth.gemspec:19]: 👍

File: lib/lp_token_auth/core.rb (hunk 3 above)
Comment (dgutt17) [reply lib/lp_token_auth/core.rb]: @bhennes2  I added some more specific error handling to the `private_key` method. Let me know what you think

**Comment by dgutt17** [comment]
@bhennes2 Ready for review again.

**Review by bhennes2 — APPROVED** [review]
LGTM!

**Pull request merged**